File: data/skills/bulk-omics-clustering/scripts/load_example_data.R
# Load example/demo data for clustering analysis testing
#
# This script loads the ALL (Acute Lymphoblastic Leukemia) dataset
# from Bioconductor for quick testing and learning.

#' Load ALL (Acute Lymphoblastic Leukemia) Microarray Data
#'
#' This function loads real patient gene expression data from the classic ALL
#' dataset (Chiaretti et al. 2004), which contains 128 pediatric ALL patients
#' with B-cell and T-cell subtypes. The data is pre-processed and suitable for
#' demonstrating clustering methods on real biological data.
#'
#' @param n_top_variable_genes Number of most variable genes to keep (default: 1000)
#'
#' @return A list containing:
#'   \item{data}{Normalized data matrix (samples × genes)}
#'   \item{metadata}{Sample annotations with cell type labels}
#'   \item{feature_names}{Gene probe IDs}
#'   \item{sample_names}{Sample identifiers}
#'   \item{true_labels}{Ground truth cell type labels (B vs T)}
#'   \item{description}{Dataset description}
#'   \item{n_samples}{Number of samples}
#'   \item{n_features}{Number of features}
#'   \item{cell_types}{Vector of cell type names}
#'
#' @details
#' - Data is log2-transformed and quantile-normalized microarray data
#' - Contains 128 samples from pediatric ALL patients
#' - Main subtypes: B-cell ALL (95 samples) vs T-cell ALL (33 samples)
#' - Runtime: ~30 seconds on first run (downloads/installs), <5 seconds cached
#' - Citation: Chiaretti et al. (2004) Blood 103(7):2771-2781
#'
#' @examples
#' \dontrun{
#' data_list <- load_example_clustering_data()
#' data <- data_list$data  # 128 samples × 1000 genes
#' metadata <- data_list$metadata
#' }
#'
#' @export
load_example_clustering_data <- function(n_top_variable_genes = 1000) {
  cat("Loading ALL (Acute Lymphoblastic Leukemia) dataset...\n")
  cat("  Source: Chiaretti et al. (2004) - 128 pediatric ALL patients\n")

  # Set CRAN mirror
  options(repos = c(CRAN = "https://cloud.r-project.org"))

  # Install BiocManager if needed
  if (!requireNamespace("BiocManager", quietly = TRUE)) {
    cat("  Installing BiocManager...\n")
    install.packages("BiocManager")
  }

  # Install ALL package if needed
  if (!requireNamespace("ALL", quietly = TRUE)) {
    cat("  Installing ALL dataset package (~5MB, ~1 min)...\n")
    BiocManager::install("ALL", update = FALSE)
  }

  # Load the dataset
  cat("  Loading data from Bioconductor...\n")
  suppressPackageStartupMessages({
    library(ALL)
    library(Biobase)
  })

  data(ALL)

  # Get expression data (genes × samples)
  expr_data <- exprs(ALL)

  # Get phenotype data
  pheno_data <- pData(ALL)

  # Extract cell type (BT = B-cell or T-cell)
  # The BT column has values like "B", "B1", "B2", "B3", "B4", "T", "T1", "T2"
  pheno_data$cell_type <- ifelse(grepl("^B", pheno_data$BT), "B", "T")

  # Transpose so samples are rows, genes are columns
  data_t <- t(expr_data)

  # Select top variable genes
  gene_vars <- apply(data_t, 2, var)
  top_genes <- names(sort(gene_vars, decreasing = TRUE)[1:n_top_variable_genes])
  data_subset <- data_t[, top_genes]

  # Z-score normalize (scale each gene)
  data_normalized <- scale(data_subset)

  # Create metadata data frame
  sample_names <- rownames(data_normalized)
  metadata <- data.frame(
    sample_id = sample_names,
    cell_type = pheno_data[sample_names, "cell_type"],
    bt_subtype = pheno_data[sample_names, "BT"],
    stringsAsFactors = FALSE
  )

  # Create numeric labels for evaluation
  true_labels <- ifelse(metadata$cell_type == "B", 0, 1)
  names(true_labels) <- sample_names

  cat("\n✓ Data loaded successfully!\n")
  cat(sprintf("  Data shape: %d samples × %d genes\n",
              nrow(data_normalized), ncol(data_normalized)))
  cat("  Data type: Microarray (log2, quantile-normalized)\n")
  cat(sprintf("  Cell types: B-cell ALL (n=%d), T-cell ALL (n=%d)\n",
              sum(metadata$cell_type == "B"),
              sum(metadata$cell_type == "T")))

  # Return list
  list(
    data = data_normalized,
    metadata = metadata,
    feature_names = colnames(data_normalized),
    sample_names = sample_names,
    true_labels = true_labels,
    description = sprintf(
      "ALL (Acute Lymphoblastic Leukemia) microarray data: 128 pediatric patients with B-cell or T-cell ALL subtypes. Data is log2-transformed, quantile-normalized, filtered to %d most variable genes. Citation: Chiaretti et al. (2004) Blood 103(7):2771-2781",
      n_top_variable_genes
    ),
    n_samples = nrow(data_normalized),
    n_features = ncol(data_normalized),
    cell_types = c("B-cell ALL", "T-cell ALL")
  )
}


#' Load a Smaller Subset of ALL Dataset for Very Quick Testing
#'
#' @return Same structure as load_example_clustering_data() but with:
#'   - 128 samples (same as full dataset)
#'   - 100 most variable genes (vs 1000 default)
#'
#' @details
#' Use this for rapid testing (<5 seconds). For learning, use the
#' default load_example_clustering_data() with 1000 genes.
#'
#' @examples
#' \dontrun{
#' small_data <- load_example_clustering_data_small()
#' }
#'
#' @export
load_example_clustering_data_small <- function() {
  load_example_clustering_data(n_top_variable_genes = 100)
}


# Test the function when sourced directly
if (sys.nframe() == 0) {
  cat(rep("=", 80), "\n", sep = "")
  cat("Testing Example Data Loader\n")
  cat(rep("=", 80), "\n", sep = "")

  # Test default parameters
  cat("\n1. Loading default example data...\n")
  data_list <- load_example_clustering_data()
  cat("\nReturned components:", paste(names(data_list), collapse = ", "), "\n")
  cat("Description:", data_list$description, "\n")

  # Test small version
  cat("\n", rep("=", 80), "\n", sep = "")
  cat("2. Loading small example data...\n")
  small_data_list <- load_example_clustering_data_small()
  cat("\nReturned components:", paste(names(small_data_list), collapse = ", "), "\n")
  cat("Description:", small_data_list$description, "\n")

  cat("\n", rep("=", 80), "\n", sep = "")
  cat("✓ All tests passed!\n")
}

File: data/skills/bulk-omics-clustering/scripts/load_example_data.py
"""
Load example/demo data for clustering analysis testing.

This module provides ALL-like synthetic data for Python users.
For the authentic ALL dataset, use the R implementation (scripts/load_example_data.R).
"""

import numpy as np
import pandas as pd
from typing import Dict, Any
import warnings


def load_example_clustering_data(
    n_top_variable_genes: int = 1000
) -> Dict[str, Any]:
    """
    Load ALL-like synthetic data for clustering analysis testing.

    This function generates synthetic data that mimics the structure of the
    ALL (Acute Lymphoblastic Leukemia) dataset: 128 samples with B-cell and
    T-cell subtypes. This is suitable for testing clustering workflows in Python.

    For the authentic ALL dataset from Bioconductor, use the R implementation:
    source("scripts/load_example_data.R")

    Parameters
    ----------
    n_top_variable_genes : int, default=1000
        Number of features (genes) to generate

    Returns
    -------
    dict
        Dictionary containing:
        - 'data': np.ndarray, normalized data matrix (samples × features)
        - 'metadata': pd.DataFrame, sample annotations with cell type labels
        - 'feature_names': list, gene probe IDs
        - 'sample_names': list, sample identifiers
        - 'description': str, dataset description
        - 'true_labels': np.ndarray, ground truth cell type labels (B vs T)

    Notes
    -----
    - Synthetic data mimicking ALL structure (95 B-cell, 33 T-cell samples)
    - Data is z-score normalized
    - Runtime: <5 seconds
    - For authentic ALL data, use R: source("scripts/load_example_data.R")

    Examples
    --------
    >>> data_dict = load_example_clustering_data()
    >>> print(data_dict['description'])
    >>> data = data_dict['data']  # Shape: (128, 1000)
    >>> metadata = data_dict['metadata']
    """
    print("Loading ALL-like synthetic data for Python testing...")
    print("  Note: For authentic ALL data from Bioconductor, use R implementation")
    print("        source('scripts/load_example_data.R')")

    data_normalized, metadata, feature_names, sample_names = _generate_all_like_data(n_top_variable_genes)

    # Create numeric labels for clustering evaluation
    cell_type_map = {'B': 0, 'T': 1}
    true_labels = np.array([cell_type_map[ct] for ct in metadata['cell_type']])

    print(f"\n✓ Data loaded successfully!")
    print(f"  Data shape: {data_normalized.shape}")
    print(f"  Data type: Synthetic data (z-score normalized)")
    print(f"  Cell types: B-cell-like (n={sum(metadata['cell_type']=='B')}), "
          f"T-cell-like (n={sum(metadata['cell_type']=='T')})")

    return {
        'data': data_normalized,
        'metadata': metadata,
        'feature_names': feature_names,
        'sample_names': sample_names,
        'true_labels': true_labels,
        'description': (
            f"ALL-like synthetic data for Python testing: "
            f"128 samples mimicking B-cell and T-cell ALL subtypes. "
            f"Data is z-score normalized with {n_top_variable_genes} features. "
            f"For authentic ALL data, use R: source('scripts/load_example_data.R')"
        ),
        'n_samples': data_normalized.shape[0],
        'n_features': data_normalized.shape[1],
        'cell_types': ['B-cell-like', 'T-cell-like']
    }


def _generate_all_like_data(n_top_variable_genes: int = 1000):
    """
    Generate synthetic data that mimics ALL dataset structure.

    This creates data similar to the ALL (Acute Lymphoblastic Leukemia) dataset
    with 95 B-cell samples and 33 T-cell samples, with distinct expression patterns.
    """
    # Generate data that mimics ALL structure
    np.random.seed(42)
    n_b_samples = 95
    n_t_samples = 33
    n_samples = n_b_samples + n_t_samples

    # Create distinct B-cell and T-cell patterns
    b_cell_data = np.random.randn(n_b_samples, n_top_variable_genes) + 0.5
    t_cell_data = np.random.randn(n_t_samples, n_top_variable_genes) - 0.5

    data = np.vstack([b_cell_data, t_cell_data])

    # Z-score normalize
    from sklearn.preprocessing import StandardScaler
    scaler = StandardScaler()
    data_normalized = scaler.fit_transform(data)

    # Create metadata
    sample_names = [f"ALL_{i:03d}" for i in range(n_samples)]
    cell_types = ['B'] * n_b_samples + ['T'] * n_t_samples
    bt_subtypes = ([f"B{i%4+1}" for i in range(n_b_samples)] +
                   [f"T{i%2+1}" for i in range(n_t_samples)])

    metadata = pd.DataFrame({
        'sample_id': sample_names,
        'cell_type': cell_types,
        'bt_subtype': bt_subtypes
    })

    feature_names = [f"Probe_{i:05d}" for i in range(n_top_variable_genes)]

    return data_normalized, metadata, feature_names, sample_names


def load_example_clustering_data_small() -> Dict[str, Any]:
    """
    Load a smaller subset of ALL-like data for very quick testing.

    Returns
    -------
    dict
        Same structure as load_example_clustering_data() but with:
        - 128 samples (same as full dataset)
        - 100 features (vs 1000 default)

    Notes
    -----
    Use this for rapid testing (<5 seconds). For learning, use the
    default load_example_clustering_data() with 1000 features.
    """
    return load_example_clustering_data(n_top_variable_genes=100)


if __name__ == "__main__":
    # Test the function
    print("=" * 80)
    print("Testing Example Data Loader")
    print("=" * 80)

    # Test default parameters
    print("\n1. Loading default example data...")
    data_dict = load_example_clustering_data()
    print(f"\nReturned keys: {list(data_dict.keys())}")
    print(f"Description: {data_dict['description']}")

    # Test small version
    print("\n" + "=" * 80)
    print("2. Loading small example data...")
    small_data_dict = load_example_clustering_data_small()
    print(f"\nReturned keys: {list(small_data_dict.keys())}")
    print(f"Description: {small_data_dict['description']}")

    print("\n" + "=" * 80)
    print("✓ All tests passed!")
